include/ignition/rendering/base/BaseVisual.hh

```diff
@@ -471,6 +475,80 @@ namespace ignition
     {
       return this->visibilityFlags;
     }
+
+    //////////////////////////////////////////////////
+    template <class T>
+    VisualPtr BaseVisual<T>::Clone(const std::string &_name,
+        NodePtr _newParent) const
+    {
+      ScenePtr scene_ = this->Scene();
+      if (nullptr == scene_)
+      {
+        ignerr << "Cloning a visual failed because the visual to be cloned is "
+          << "not attached to a scene.\n";
+        return nullptr;
+      }
+      VisualPtr result;
+      if (_name.empty())
+        result = scene_->CreateVisual();
+      else
+        result = scene_->CreateVisual(_name);
+
+      if (nullptr != _newParent)
+      {
+        auto parentScene = _newParent->Scene();
+        if (nullptr != parentScene && parentScene->Id() != scene_->Id())
+        {
+          ignerr << "Cloning a visual failed because the desired parent of the "
+            << "cloned visual belongs to a different scene.\n";
+          scene_->DestroyVisual(result);
+          return nullptr;
+        }
+        _newParent->AddChild(result);
+      }
+
+      result->SetOrigin(this->Origin());
+      result->SetInheritScale(this->InheritScale());
+      result->SetLocalScale(this->LocalScale());
+      result->SetLocalPose(this->LocalPose());
+      result->SetVisibilityFlags(this->VisibilityFlags());
+      result->SetWireframe(this->Wireframe());
+
+      // if the visual that was cloned has child visuals, clone those as well
+      auto children_ =
+          std::dynamic_pointer_cast<BaseStore<ignition::rendering::Node, T>>(
+          this->Children());
+      if (!children_)
+      {
+        ignerr << "Cast failed in BaseVisual::Clone\n";
+        scene_->DestroyVisual(result);
+        return nullptr;
+      }
+      for (auto it = children_->Begin(); it != children_->End(); ++it)
+      {
+        NodePtr child = it->second;
+        VisualPtr visual = std::dynamic_pointer_cast<Visual>(child);
+        if (nullptr != visual)
+          visual->Clone("", result);
+      }
+
+      for (unsigned int i = 0; i < this->GeometryCount(); ++i)
+        result->AddGeometry(this->GeometryByIndex(i)->Clone());
+
+      if (this->Material())
+        result->SetMaterial(this->Material());
+
+      // should userData be cloned? Or should that be left out since (I believe)
+      // userData is intended to be custom/unique for each visual
+      //
+      // copying userData seems to segfault, so this is commented out for now
+      /*
+       *for (const auto &[key, val] : this->userData)
+       *  result->SetUserData(key, val);
+       */
```

Comment:
I'm not sure if `userData` should be cloned, but when I tried to clone visuals, the for loop that copies the `userData` to the clone was causing a segfault

Reply:
this copies data such as temperature and laser retro associated with a visual, which could be useful where there are thermal cameras / lidars in the scene. Do you know what's causing the segfault?

Reply:
Good point about things like temperature and laser retro. The segfaults I was seeing were when I was using this with https://github.com/ignitionrobotics/ign-gazebo/pull/1013 to copy/paste entities, but now, I am not seeing that segfault anymore in my most recent tests. So, the issue must have been something else that I fixed recently, which means that we can clone `userData` (see https://github.com/ignitionrobotics/ign-rendering/pull/397/commits/0a3f104a324b6d5871c600c269a55a7928395b5b).